clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3524,44 +3524,62 @@ void handleResponse(AbstractResponse abstractResponse) {
                 for (final Node node : allNodes) {
                     final long nowList = time.milliseconds();
                     runnable.call(new Call("listGroups", deadline, new ConstantNodeIdProvider(node.id())) {
+
+                        // If only regular consumer group types are required, we can try an earlier request version if
+                        // UnsupportedVersionException is thrown
+                        final boolean canTryEarlierRequestVersion = options.regularConsumerGroupTypes();
```

Comment:
Instead of re-sending the request, maybe we can modify the request (ListGroupsRequest.Builder#build) when the version is small than v5?
```java
            if (!data.typesFilter().isEmpty() && version < 5) {
                var copy = new HashSet<>(data.typesFilter());
                copy.remove(GroupType.CLASSIC.toString());
                copy.remove(GroupType.CONSUMER.toString());
                if (!copy.isEmpty())
                    throw new UnsupportedVersionException("The broker only supports ListGroups v" + version + ", but we need v5 or newer to request groups by type.");
                return new ListGroupsRequest(data.duplicate().setTypesFilter(List.of()), version);
            }
```

Reply:
I've taken this suggestion and tweaked it a bit.